src/Domain/CoordinateMaps/TimeDependent/ShapeMapTransitionFunctions/Wedge.cpp

```diff
@@ -5,139 +5,416 @@
+      for (size_t i = 0; i < get_size(result); i++) {
+        if (get_element(gsl::at(centered_coords, axis_idx), i) == 0.0) {
+          get_element(result, i) = -std::numeric_limits<double>::max();
+        } else {
+          get_element(result, i) /=
+              get_element(gsl::at(centered_coords, axis_idx), i);
+        }
+      }
+    }
+
+    return result;
+  };
+
+  // If we specified an axis that we are on, then use that lambda. This saves
+  // some computation.
+  if (potential_axis.has_value()) {
+    return lambda_axis(potential_axis.value());
+  }
+
+  // Otherwise we need to loop over all of them
+  const std::array all_axes{Axis::PlusZ,  Axis::MinusZ, Axis::PlusY,
+                            Axis::MinusY, Axis::PlusX,  Axis::MinusX};
+  std::array<T, 6> candidate_lambdas{};
+
+  for (size_t i = 0; i < all_axes.size(); i++) {
+    gsl::at(candidate_lambdas, i) = lambda_axis(gsl::at(all_axes, i));
+  }
+
+  // We need to take the min of all the positive lambdas
+  const double fill_value = std::numeric_limits<double>::max();
+  T result = make_with_value<T>(candidate_lambdas[0], fill_value);
+  std::vector<Axis> chosen_axes(get_size(result), Axis::None);
+  for (size_t i = 0; i < get_size(result); i++) {
+    for (size_t j = 0; j < candidate_lambdas.size(); j++) {
+      if (get_element(gsl::at(candidate_lambdas, j), i) > 0.0 and
+          get_element(gsl::at(candidate_lambdas, j), i) <
+              get_element(result, i)) {
+        get_element(result, i) = get_element(gsl::at(candidate_lambdas, j), i);
+        chosen_axes[i] = gsl::at(all_axes, j);
+      }
+    }
+  }
+
+  ASSERT(not alg::any_of(chosen_axes,
+                         [&](const Axis& a) { return a == Axis::None; }),
+         "No candidate cube lambdas in the Wedge shape map transition function "
+         "are negative. This is an internal error. Please file an issue.");
+
+  return result;
 }
 
-bool Wedge::Surface::operator!=(const Wedge::Surface& other) const {
-  return not(*this == other);
+template <typename T>
+T Wedge::lambda_sphere(const std::array<T, 3>& centered_coords,
+                       const T& centered_coords_magnitude) const {
+  // quadratic equation is
+  // a lambda^2 + b lambda + c = 0
+  //
+  // For this case
+  // a = |x^i-P^i|^2
+  // b = 2(x^i-P^i)(P^j-C^j)\delta_{ij}
+  // c = |P^i-C^i|^2 - R^2
+
+  const T a = square(centered_coords_magnitude);
+  const T b = 2.0 * dot(centered_coords, projection_center_);
+  const T c =
+      make_with_value<T>(b, dot(projection_center_, projection_center_) -
+                                square(outer_surface_.radius));
+
+  CAPTURE_FOR_ERROR(a);
+  CAPTURE_FOR_ERROR(b);
+  CAPTURE_FOR_ERROR(c);
+
+  // The root that we are looking for is positive (since a negative root would
+  // be for the opposite side of the sphere)
+  return smallest_root_greater_than_value_within_roundoff(
+      a, b, c, -std::numeric_limits<double>::epsilon() * 100.0);
```

Comment:
Could you document why this epsilon? I assume it's to handle effectively zero but slightly negative?